Fix removing vendor folders (FAQ 1.44)

williamdes opened the pull request:
### Description

Ref: https://github.com/Roave/SecurityAdvisories/issues/76
Fixes: https://github.com/phpmyadmin/phpmyadmin/issues/17661
Ref: #16418
Fixes: #16419

williamdes committed:
Add a security to check if all files present in the autoload class map can be found and re-generate the class map

Signed-off-by: William Desportes <williamdes@wdes.fr>

williamdes committed:
Update FAQ 1.44

Signed-off-by: William Desportes <williamdes@wdes.fr>

williamdes commented on composer.json:
```diff
@@ -24,7 +24,19 @@
+        "exclude-from-classmap": [
+            "/test/",
+            "/vendor/williamdes/mariadb-mysql-kbs",
+            "/vendor/tecnickcom/tcpdf/tcpdf_barcodes_*.php",
+            "/vendor/tecnickcom/tcpdf/tcpdf_import.php",
+            "/vendor/tecnickcom/tcpdf/tcpdf_parser.php",
+            "/vendor/tecnickcom/tcpdf/include/tcpdf_filters.php",
+            "/vendor/tecnickcom/tcpdf/include/barcodes",
```
just a cleanup, we do not need the classes do be in the class map as we do not use them

MauricioFauth replied:
Do we really have to maintain this list of excludes?

williamdes replied:
Yes, unfortunately. It's not really needed about the tcpdf lib, I can remove the lines. At some point I would like to add the files to the vendor exclude removal list. But for now each time I do this composer fails building the class map because they are hard coded files in tcpdf's composer json and if files are missing it gets un happy

williamdes commented on composer.json:
```diff
@@ -24,7 +24,19 @@
-        "files": ["vendor/phpmyadmin/motranslator/src/functions.php"]
+        "files": ["vendor/phpmyadmin/motranslator/src/functions.php"],
+        "exclude-from-classmap": [
+            "/test/",
```
The files are present when we do composer commands in the release script so so it dumps class maps for it, that's not problematic but very useless and false.
It also breaks my code I wrote to check all classmap files exist

MauricioFauth commented on doc/faq.rst:
```diff
@@ -690,12 +690,17 @@ A list of files and corresponding functionality which degrade gracefully when re
+* :file:`./js/config/` (Configuration files to re-build `./js/dist/`)
+* :file:`./vendor/williamdes/mariadb-mysql-kbs/` folder (external links to MariaDB and MySQL documentations)
+* :file:`./vendor/code-lts/u2f-php-server/` folder (U2F second factor authentication)
+* :file:`./vendor/pragmarx/google2fa-qrcode/` folder (2FA second factor authentication)
```
I'm not sure about this. Instead of adding more vendor files to remove, I think we should remove the existing ones.

williamdes replied:
Hmm, no ?
Why should we remove our vendors that make 2FA possible ?

Or do you mean saying that no vendor is removable ?

MauricioFauth replied:
Any vendor directory. That's not the proper way to "uninstall" a dependency.

williamdes replied:
That's true, I will change the documentation about it

williamdes committed:
Add the version to composer.json at release build time

Ref: https://github.com/Roave/SecurityAdvisories/issues/76
Fixes: https://github.com/phpmyadmin/phpmyadmin/issues/17661

Signed-off-by: William Desportes <williamdes@wdes.fr>

williamdes commented on doc/faq.rst:
```diff
@@ -686,16 +686,19 @@ Some users have requested to be able to reduce the size of the phpMyAdmin instal
+* :file:`./sql/` (SQL scripts to configure advanced functionalities)
+* :file:`./js/src/` (Source files to re-build `./js/dist/`)
+* :file:`./js/config/` (Configuration files to re-build `./js/dist/`)
+* Run `composer remove --update-no-dev tecnickcom/tcpdf` (exporting to PDF)
+* Run `composer remove --update-no-dev williamdes/mariadb-mysql-kbs` (external links to MariaDB and MySQL documentations)
+* Run `composer remove --update-no-dev code-lts/u2f-php-server` (U2F second factor authentication)
+* Run `composer remove --update-no-dev pragmarx/google2fa-qrcode` (2FA second factor authentication)
```
@MauricioFauth this is problematic because it will remove all other vendors we installed manually from the suggest space. But because we did revert the changes to composer.json composer does not know why they are here and removes them

```
phpMyAdmin-5.2.1-dev-all-languages$ composer remove williamdes/mariadb-mysql-kbs --dry-run --update-no-dev --no-interaction
williamdes/mariadb-mysql-kbs is not required in your composer.json and has not been removed
./composer.json has been updated
Running composer update williamdes/mariadb-mysql-kbs
Loading composer repositories with package information
Package "williamdes/mariadb-mysql-kbs" listed for update is not locked.
Updating dependencies
Nothing to modify in lock file
Installing dependencies from lock file
Package operations: 0 installs, 0 updates, 8 removals
  - Removing williamdes/mariadb-mysql-kbs (v1.2.13)
  - Removing tecnickcom/tcpdf (6.5.0)
  - Removing pragmarx/google2fa-qrcode (v2.1.1)
  - Removing pragmarx/google2fa (v8.0.1)
  - Removing paragonie/constant_time_encoding (v2.6.3)
  - Removing dasprid/enum (1.0.3)
  - Removing code-lts/u2f-php-server (v1.2.0)
  - Removing bacon/bacon-qr-code (2.0.7)
```

MauricioFauth replied:
What about using `--no-install` instead of `--update-no-dev`?

williamdes replied:
Well the `--no-install` will make that the package is not removed and composer will warn about it. Basically it does nothing

MauricioFauth replied:
What about adding the optional dependencies that are included in the release to the composer.json's `require` with the release script?

williamdes replied:
Yes, it's an option, but anyone doing packaging will then be blocked because of that. The libs will be required but they still are not. Maybe we could ask composer team some help ?

MauricioFauth commented on scripts/create-release.sh:
```diff
@@ -468,15 +509,27 @@ do
-echo "Installing composer packages '$PACKAGES_VERSIONS'"
+echo "* Installing composer packages '$PACKAGES_VERSIONS'"
 
-composer require --no-interaction --optimize-autoloader --update-no-dev $PACKAGES_VERSIONS
```
Why removing `--optimize-autoloader` was necessary?

MauricioFauth commented on scripts/create-release.sh:
```diff
@@ -284,7 +284,7 @@ delete_phpunit_sandbox() {
         exit 1;
     fi
-    rm -rf "${TEMP_PHPUNIT_FOLDER}"
+    rm -r "${TEMP_PHPUNIT_FOLDER}"
```
Not related to this PR, but could we use the [PHPUnit's PHAR](https://phpunit.de/getting-started/phpunit-9.html) instead of installing it in a sandbox?

williamdes replied:
Yeah probably if we can extract the version that we use. It's probably more complicated than the sandbox code. Or maybe not

williamdes commented on composer.json:
```diff
@@ -24,7 +24,19 @@
+            "/vendor/tecnickcom/tcpdf/include/barcodes",
+            "/vendor/pragmarx/google2fa-qrcode",
+            "/vendor/bacon/bacon-qr-code",
+            "/vendor/code-lts/u2f-php-server"
```
Other excludes could be removed if we solve the composer remove library problem

williamdes committed:
Remove tcpdf autoloaded files

Signed-off-by: William Desportes <williamdes@wdes.fr>

williamdes committed:
Fix CI version to have a valid composer.json

```
Failed to update composer.json with a valid format, reverting to the original content. Please report an issue to us with details (command you run and a copy of your composer.json).
  version : Does not match the regex pattern ^v?\d+(\.\d+){0,3}|^dev-
```

Signed-off-by: William Desportes <williamdes@wdes.fr>

williamdes committed:
Remove non needed yarn install on build release step

modules are not needed to launch the create release script

Signed-off-by: William Desportes <williamdes@wdes.fr>

MauricioFauth commented on .github/workflows/other-tools.yml:
```diff
@@ -70,9 +70,6 @@ jobs:
           restore-keys: |
             ${{ runner.os }}-yarn-
 
-      - name: Install modules
-        run: yarn install --non-interactive
```
Shouldn't the previous step that caches the Yarn dependencies be removed as well?

williamdes replied:
Probably not because the install normally uses the system cache

MauricioFauth replied:
:100:
